File: src/main/java/com/merge/api/resources/filestorage/types/Folder.java
package com.merge.api.resources.filestorage.types;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonSetter;
import com.fasterxml.jackson.annotation.Nulls;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import java.time.OffsetDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;

@JsonInclude(JsonInclude.Include.NON_EMPTY)
@JsonDeserialize(builder = Folder.Builder.class)
public final class Folder {
    private final Optional<String> id;

    private final Optional<String> remoteId;

    private final Optional<String> name;

    private final Optional<String> folderUrl;

    private final Optional<Integer> size;

    private final Optional<String> description;

    private final Optional<String> parentFolder;

    private final Optional<String> drive;

    private final List<String> permissions;

    private final Optional<OffsetDateTime> remoteCreatedAt;

    private final Optional<OffsetDateTime> remoteUpdatedAt;

    private final Optional<Boolean> remoteWasDeleted;

    private final Optional<OffsetDateTime> modifiedAt;

    private final Optional<Map<String, JsonNode>> fieldMappings;

    private final Optional<List<Optional<Map<String, JsonNode>>>> remoteData;

    private Folder(
            Optional<String> id,
            Optional<String> remoteId,
            Optional<String> name,
            Optional<String> folderUrl,
            Optional<Integer> size,
            Optional<String> description,
            Optional<String> parentFolder,
            Optional<String> drive,
            List<String> permissions,
            Optional<OffsetDateTime> remoteCreatedAt,
            Optional<OffsetDateTime> remoteUpdatedAt,
            Optional<Boolean> remoteWasDeleted,
            Optional<OffsetDateTime> modifiedAt,
            Optional<Map<String, JsonNode>> fieldMappings,
            Optional<List<Optional<Map<String, JsonNode>>>> remoteData) {
        this.id = id;
        this.remoteId = remoteId;
        this.name = name;
        this.folderUrl = folderUrl;
        this.size = size;
        this.description = description;
        this.parentFolder = parentFolder;
        this.drive = drive;
        this.permissions = permissions;
        this.remoteCreatedAt = remoteCreatedAt;
        this.remoteUpdatedAt = remoteUpdatedAt;
        this.remoteWasDeleted = remoteWasDeleted;
        this.modifiedAt = modifiedAt;
        this.fieldMappings = fieldMappings;
        this.remoteData = remoteData;
    }

    @JsonProperty("id")
    public Optional<String> getId() {
        return id;
    }

    /**
     * @return The third-party API ID of the matching object.
     */
    @JsonProperty("remote_id")
    public Optional<String> getRemoteId() {
        return remoteId;
    }

    /**
     * @return The folder's name.
     */
    @JsonProperty("name")
    public Optional<String> getName() {
        return name;
    }

    /**
     * @return The URL to access the folder.
     */
    @JsonProperty("folder_url")
    public Optional<String> getFolderUrl() {
        return folderUrl;
    }

    /**
     * @return The folder's size, in bytes.
     */
    @JsonProperty("size")
    public Optional<Integer> getSize() {
        return size;
    }

    /**
     * @return The folder's description.
     */
    @JsonProperty("description")
    public Optional<String> getDescription() {
        return description;
    }

    /**
     * @return The folder that the folder belongs to.
     */
    @JsonProperty("parent_folder")
    public Optional<String> getParentFolder() {
        return parentFolder;
    }

    /**
     * @return The drive that the folder belongs to.
     */
    @JsonProperty("drive")
    public Optional<String> getDrive() {
        return drive;
    }

    /**
     * @return The Permission object is used to represent a user's or group's access to a File or Folder. Permissions are unexpanded by default. Use the query param <code>expand=permissions</code> to see more details under <code>GET /folders</code>.
     */
    @JsonProperty("permissions")
    public List<String> getPermissions() {
        return permissions;
    }

    /**
     * @return When the third party's folder was created.
     */
    @JsonProperty("remote_created_at")
    public Optional<OffsetDateTime> getRemoteCreatedAt() {
        return remoteCreatedAt;
    }

    /**
     * @return When the third party's folder was updated.
     */
    @JsonProperty("remote_updated_at")
    public Optional<OffsetDateTime> getRemoteUpdatedAt() {
        return remoteUpdatedAt;
    }

    /**
     * @return Indicates whether or not this object has been deleted by third party webhooks.
     */
    @JsonProperty("remote_was_deleted")
    public Optional<Boolean> getRemoteWasDeleted() {
        return remoteWasDeleted;
    }

    /**
     * @return This is the datetime that this object was last updated by Merge
     */
    @JsonProperty("modified_at")
    public Optional<OffsetDateTime> getModifiedAt() {
        return modifiedAt;
    }

    @JsonProperty("field_mappings")
    public Optional<Map<String, JsonNode>> getFieldMappings() {
        return fieldMappings;
    }

    @JsonProperty("remote_data")
    public Optional<List<Optional<Map<String, JsonNode>>>> getRemoteData() {
        return remoteData;
    }

    @Override
    public boolean equals(Object other) {
        if (this == other) return true;
        return other instanceof Folder && equalTo((Folder) other);
    }

    private boolean equalTo(Folder other) {
        return id.equals(other.id)
                && remoteId.equals(other.remoteId)
                && name.equals(other.name)
                && folderUrl.equals(other.folderUrl)
                && size.equals(other.size)
                && description.equals(other.description)
                && parentFolder.equals(other.parentFolder)
                && drive.equals(other.drive)
                && permissions.equals(other.permissions)
                && remoteCreatedAt.equals(other.remoteCreatedAt)
                && remoteUpdatedAt.equals(other.remoteUpdatedAt)
                && remoteWasDeleted.equals(other.remoteWasDeleted)
                && modifiedAt.equals(other.modifiedAt)
                && fieldMappings.equals(other.fieldMappings)
                && remoteData.equals(other.remoteData);
    }

    @Override
    public int hashCode() {
        return Objects.hash(
                this.id,
                this.remoteId,
                this.name,
                this.folderUrl,
                this.size,
                this.description,
                this.parentFolder,
                this.drive,
                this.permissions,
                this.remoteCreatedAt,
                this.remoteUpdatedAt,
                this.remoteWasDeleted,
                this.modifiedAt,
                this.fieldMappings,
                this.remoteData);
    }

    @Override
    public String toString() {
        return "Folder{" + "id: " + id + ", remoteId: " + remoteId + ", name: " + name + ", folderUrl: " + folderUrl
                + ", size: " + size + ", description: " + description + ", parentFolder: " + parentFolder + ", drive: "
                + drive + ", permissions: " + permissions + ", remoteCreatedAt: " + remoteCreatedAt
                + ", remoteUpdatedAt: " + remoteUpdatedAt + ", remoteWasDeleted: " + remoteWasDeleted + ", modifiedAt: "
                + modifiedAt + ", fieldMappings: " + fieldMappings + ", remoteData: " + remoteData + "}";
    }

    public static Builder builder() {
        return new Builder();
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    public static final class Builder {
        private Optional<String> id = Optional.empty();

        private Optional<String> remoteId = Optional.empty();

        private Optional<String> name = Optional.empty();

        private Optional<String> folderUrl = Optional.empty();

        private Optional<Integer> size = Optional.empty();

        private Optional<String> description = Optional.empty();

        private Optional<String> parentFolder = Optional.empty();

        private Optional<String> drive = Optional.empty();

        private List<String> permissions = new ArrayList<>();

        private Optional<OffsetDateTime> remoteCreatedAt = Optional.empty();

        private Optional<OffsetDateTime> remoteUpdatedAt = Optional.empty();

        private Optional<Boolean> remoteWasDeleted = Optional.empty();

        private Optional<OffsetDateTime> modifiedAt = Optional.empty();

        private Optional<Map<String, JsonNode>> fieldMappings = Optional.empty();

        private Optional<List<Optional<Map<String, JsonNode>>>> remoteData = Optional.empty();

        private Builder() {}

        public Builder from(Folder other) {
            id(other.getId());
            remoteId(other.getRemoteId());
            name(other.getName());
            folderUrl(other.getFolderUrl());
            size(other.getSize());
            description(other.getDescription());
            parentFolder(other.getParentFolder());
            drive(other.getDrive());
            permissions(other.getPermissions());
            remoteCreatedAt(other.getRemoteCreatedAt());
            remoteUpdatedAt(other.getRemoteUpdatedAt());
            remoteWasDeleted(other.getRemoteWasDeleted());
            modifiedAt(other.getModifiedAt());
            fieldMappings(other.getFieldMappings());
            remoteData(other.getRemoteData());
            return this;
        }

        @JsonSetter(value = "id", nulls = Nulls.SKIP)
        public Builder id(Optional<String> id) {
            this.id = id;
            return this;
        }

        public Builder id(String id) {
            this.id = Optional.of(id);
            return this;
        }

        @JsonSetter(value = "remote_id", nulls = Nulls.SKIP)
        public Builder remoteId(Optional<String> remoteId) {
            this.remoteId = remoteId;
            return this;
        }

        public Builder remoteId(String remoteId) {
            this.remoteId = Optional.of(remoteId);
            return this;
        }

        @JsonSetter(value = "name", nulls = Nulls.SKIP)
        public Builder name(Optional<String> name) {
            this.name = name;
            return this;
        }

        public Builder name(String name) {
            this.name = Optional.of(name);
            return this;
        }

        @JsonSetter(value = "folder_url", nulls = Nulls.SKIP)
        public Builder folderUrl(Optional<String> folderUrl) {
            this.folderUrl = folderUrl;
            return this;
        }

        public Builder folderUrl(String folderUrl) {
            this.folderUrl = Optional.of(folderUrl);
            return this;
        }

        @JsonSetter(value = "size", nulls = Nulls.SKIP)
        public Builder size(Optional<Integer> size) {
            this.size = size;
            return this;
        }

        public Builder size(Integer size) {
            this.size = Optional.of(size);
            return this;
        }

        @JsonSetter(value = "description", nulls = Nulls.SKIP)
        public Builder description(Optional<String> description) {
            this.description = description;
            return this;
        }

        public Builder description(String description) {
            this.description = Optional.of(description);
            return this;
        }

        @JsonSetter(value = "parent_folder", nulls = Nulls.SKIP)
        public Builder parentFolder(Optional<String> parentFolder) {
            this.parentFolder = parentFolder;
            return this;
        }

        public Builder parentFolder(String parentFolder) {
            this.parentFolder = Optional.of(parentFolder);
            return this;
        }

        @JsonSetter(value = "drive", nulls = Nulls.SKIP)
        public Builder drive(Optional<String> drive) {
            this.drive = drive;
            return this;
        }

        public Builder drive(String drive) {
            this.drive = Optional.of(drive);
            return this;
        }

        @JsonSetter(value = "permissions", nulls = Nulls.SKIP)
        public Builder permissions(List<String> permissions) {
            this.permissions.clear();
            this.permissions.addAll(permissions);
            return this;
        }

        public Builder addPermissions(String permissions) {
            this.permissions.add(permissions);
            return this;
        }

        public Builder addAllPermissions(List<String> permissions) {
            this.permissions.addAll(permissions);
            return this;
        }

        @JsonSetter(value = "remote_created_at", nulls = Nulls.SKIP)
        public Builder remoteCreatedAt(Optional<OffsetDateTime> remoteCreatedAt) {
            this.remoteCreatedAt = remoteCreatedAt;
            return this;
        }

        public Builder remoteCreatedAt(OffsetDateTime remoteCreatedAt) {
            this.remoteCreatedAt = Optional.of(remoteCreatedAt);
            return this;
        }

        @JsonSetter(value = "remote_updated_at", nulls = Nulls.SKIP)
        public Builder remoteUpdatedAt(Optional<OffsetDateTime> remoteUpdatedAt) {
            this.remoteUpdatedAt = remoteUpdatedAt;
            return this;
        }

        public Builder remoteUpdatedAt(OffsetDateTime remoteUpdatedAt) {
            this.remoteUpdatedAt = Optional.of(remoteUpdatedAt);
            return this;
        }

        @JsonSetter(value = "remote_was_deleted", nulls = Nulls.SKIP)
        public Builder remoteWasDeleted(Optional<Boolean> remoteWasDeleted) {
            this.remoteWasDeleted = remoteWasDeleted;
            return this;
        }

        public Builder remoteWasDeleted(Boolean remoteWasDeleted) {
            this.remoteWasDeleted = Optional.of(remoteWasDeleted);
            return this;
        }

        @JsonSetter(value = "modified_at", nulls = Nulls.SKIP)
        public Builder modifiedAt(Optional<OffsetDateTime> modifiedAt) {
            this.modifiedAt = modifiedAt;
            return this;
        }

        public Builder modifiedAt(OffsetDateTime modifiedAt) {
            this.modifiedAt = Optional.of(modifiedAt);
            return this;
        }

        @JsonSetter(value = "field_mappings", nulls = Nulls.SKIP)
        public Builder fieldMappings(Optional<Map<String, JsonNode>> fieldMappings) {
            this.fieldMappings = fieldMappings;
            return this;
        }

        public Builder fieldMappings(Map<String, JsonNode> fieldMappings) {
            this.fieldMappings = Optional.of(fieldMappings);
            return this;
        }

        @JsonSetter(value = "remote_data", nulls = Nulls.SKIP)
        public Builder remoteData(Optional<List<Optional<Map<String, JsonNode>>>> remoteData) {
            this.remoteData = remoteData;
            return this;
        }

        public Builder remoteData(List<Optional<Map<String, JsonNode>>> remoteData) {
            this.remoteData = Optional.of(remoteData);
            return this;
        }

        public Folder build() {
            return new Folder(
                    id,
                    remoteId,
                    name,
                    folderUrl,
                    size,
                    description,
                    parentFolder,
                    drive,
                    permissions,
                    remoteCreatedAt,
                    remoteUpdatedAt,
                    remoteWasDeleted,
                    modifiedAt,
                    fieldMappings,
                    remoteData);
        }
    }
}

File: merge-java-core/src/main/kotlin/dev/merge/api/models/fileStorage/Folder.kt
package dev.merge.api.models.fileStorage

import com.fasterxml.jackson.annotation.JsonAnyGetter
import com.fasterxml.jackson.annotation.JsonAnySetter
import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import dev.merge.api.core.ExcludeMissing
import dev.merge.api.core.JsonField
import dev.merge.api.core.JsonMissing
import dev.merge.api.core.JsonValue
import dev.merge.api.core.NoAutoDetect
import dev.merge.api.core.toUnmodifiable
import dev.merge.api.models.*
import java.time.OffsetDateTime
import java.util.Objects
import java.util.Optional

/**
 * # The Folder Object
 *
 * ### Description
 *
 * The `Folder` object is used to represent a collection of files and/or folders in the workspace.
 * Could be within a drive, if it exsts.
 *
 * ### Usage Example
 *
 * Fetch from the `GET /api/filestorage/v1/folders` endpoint and view their folders.
 */
@JsonDeserialize(builder = Folder.Builder::class)
@NoAutoDetect
class Folder
private constructor(
    private val id: JsonField<String>,
    private val remoteId: JsonField<String>,
    private val name: JsonField<String>,
    private val folderUrl: JsonField<String>,
    private val size: JsonField<Long>,
    private val description: JsonField<String>,
    private val parentFolder: JsonField<String>,
    private val drive: JsonField<String>,
    private val permissions: JsonField<List<String>>,
    private val remoteCreatedAt: JsonField<OffsetDateTime>,
    private val remoteUpdatedAt: JsonField<OffsetDateTime>,
    private val fieldMappings: JsonValue,
    private val modifiedAt: JsonField<OffsetDateTime>,
    private val remoteData: JsonField<List<JsonValue>>,
    private val additionalProperties: Map<String, JsonValue>,
) {

    private var validated: Boolean = false

    private var hashCode: Int = 0

    fun id(): Optional<String> = Optional.ofNullable(id.getNullable("id"))

    /** The third-party API ID of the matching object. */
    fun remoteId(): Optional<String> = Optional.ofNullable(remoteId.getNullable("remote_id"))

    /** The folder's name. */
    fun name(): Optional<String> = Optional.ofNullable(name.getNullable("name"))

    /** The URL to access the folder. */
    fun folderUrl(): Optional<String> = Optional.ofNullable(folderUrl.getNullable("folder_url"))

    /** The folder's size, in bytes. */
    fun size(): Optional<Long> = Optional.ofNullable(size.getNullable("size"))

    /** The folder's description. */
    fun description(): Optional<String> =
        Optional.ofNullable(description.getNullable("description"))

    /** The folder that the folder belongs to. */
    fun parentFolder(): Optional<String> =
        Optional.ofNullable(parentFolder.getNullable("parent_folder"))

    /** The drive that the folder belongs to. */
    fun drive(): Optional<String> = Optional.ofNullable(drive.getNullable("drive"))

    /**
     * The Permission object is used to represent a user's or group's access to a File or Folder.
     * Permissions are unexpanded by default. Use the query param `expand=permissions` to see more
     * details under `GET /folders`.
     */
    fun permissions(): List<String> = permissions.getRequired("permissions")

    /** When the third party's folder was created. */
    fun remoteCreatedAt(): Optional<OffsetDateTime> =
        Optional.ofNullable(remoteCreatedAt.getNullable("remote_created_at"))

    /** When the third party's folder was updated. */
    fun remoteUpdatedAt(): Optional<OffsetDateTime> =
        Optional.ofNullable(remoteUpdatedAt.getNullable("remote_updated_at"))

    /** This is the datetime that this object was last updated by Merge */
    fun modifiedAt(): Optional<OffsetDateTime> =
        Optional.ofNullable(modifiedAt.getNullable("modified_at"))

    fun remoteData(): Optional<List<JsonValue>> =
        Optional.ofNullable(remoteData.getNullable("remote_data"))

    @JsonProperty("id") @ExcludeMissing fun _id() = id

    /** The third-party API ID of the matching object. */
    @JsonProperty("remote_id") @ExcludeMissing fun _remoteId() = remoteId

    /** The folder's name. */
    @JsonProperty("name") @ExcludeMissing fun _name() = name

    /** The URL to access the folder. */
    @JsonProperty("folder_url") @ExcludeMissing fun _folderUrl() = folderUrl

    /** The folder's size, in bytes. */
    @JsonProperty("size") @ExcludeMissing fun _size() = size

    /** The folder's description. */
    @JsonProperty("description") @ExcludeMissing fun _description() = description

    /** The folder that the folder belongs to. */
    @JsonProperty("parent_folder") @ExcludeMissing fun _parentFolder() = parentFolder

    /** The drive that the folder belongs to. */
    @JsonProperty("drive") @ExcludeMissing fun _drive() = drive

    /**
     * The Permission object is used to represent a user's or group's access to a File or Folder.
     * Permissions are unexpanded by default. Use the query param `expand=permissions` to see more
     * details under `GET /folders`.
     */
    @JsonProperty("permissions") @ExcludeMissing fun _permissions() = permissions

    /** When the third party's folder was created. */
    @JsonProperty("remote_created_at") @ExcludeMissing fun _remoteCreatedAt() = remoteCreatedAt

    /** When the third party's folder was updated. */
    @JsonProperty("remote_updated_at") @ExcludeMissing fun _remoteUpdatedAt() = remoteUpdatedAt

    @JsonProperty("field_mappings") @ExcludeMissing fun _fieldMappings() = fieldMappings

    /** This is the datetime that this object was last updated by Merge */
    @JsonProperty("modified_at") @ExcludeMissing fun _modifiedAt() = modifiedAt

    @JsonProperty("remote_data") @ExcludeMissing fun _remoteData() = remoteData

    @JsonAnyGetter
    @ExcludeMissing
    fun _additionalProperties(): Map<String, JsonValue> = additionalProperties

    fun validate() = apply {
        if (!validated) {
            id()
            remoteId()
            name()
            folderUrl()
            size()
            description()
            parentFolder()
            drive()
            permissions()
            remoteCreatedAt()
            remoteUpdatedAt()
            modifiedAt()
            remoteData()
            validated = true
        }
    }

    fun toBuilder() = Builder().from(this)

    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        }

        return other is Folder &&
            this.id == other.id &&
            this.remoteId == other.remoteId &&
            this.name == other.name &&
            this.folderUrl == other.folderUrl &&
            this.size == other.size &&
            this.description == other.description &&
            this.parentFolder == other.parentFolder &&
            this.drive == other.drive &&
            this.permissions == other.permissions &&
            this.remoteCreatedAt == other.remoteCreatedAt &&
            this.remoteUpdatedAt == other.remoteUpdatedAt &&
            this.fieldMappings == other.fieldMappings &&
            this.modifiedAt == other.modifiedAt &&
            this.remoteData == other.remoteData &&
            this.additionalProperties == other.additionalProperties
    }

    override fun hashCode(): Int {
        if (hashCode == 0) {
            hashCode =
                Objects.hash(
                    id,
                    remoteId,
                    name,
                    folderUrl,
                    size,
                    description,
                    parentFolder,
                    drive,
                    permissions,
                    remoteCreatedAt,
                    remoteUpdatedAt,
                    fieldMappings,
                    modifiedAt,
                    remoteData,
                    additionalProperties,
                )
        }
        return hashCode
    }

    override fun toString() =
        "Folder{id=$id, remoteId=$remoteId, name=$name, folderUrl=$folderUrl, size=$size, description=$description, parentFolder=$parentFolder, drive=$drive, permissions=$permissions, remoteCreatedAt=$remoteCreatedAt, remoteUpdatedAt=$remoteUpdatedAt, fieldMappings=$fieldMappings, modifiedAt=$modifiedAt, remoteData=$remoteData, additionalProperties=$additionalProperties}"

    companion object {

        @JvmStatic fun builder() = Builder()
    }

    class Builder {

        private var id: JsonField<String> = JsonMissing.of()
        private var remoteId: JsonField<String> = JsonMissing.of()
        private var name: JsonField<String> = JsonMissing.of()
        private var folderUrl: JsonField<String> = JsonMissing.of()
        private var size: JsonField<Long> = JsonMissing.of()
        private var description: JsonField<String> = JsonMissing.of()
        private var parentFolder: JsonField<String> = JsonMissing.of()
        private var drive: JsonField<String> = JsonMissing.of()
        private var permissions: JsonField<List<String>> = JsonMissing.of()
        private var remoteCreatedAt: JsonField<OffsetDateTime> = JsonMissing.of()
        private var remoteUpdatedAt: JsonField<OffsetDateTime> = JsonMissing.of()
        private var fieldMappings: JsonValue = JsonMissing.of()
        private var modifiedAt: JsonField<OffsetDateTime> = JsonMissing.of()
        private var remoteData: JsonField<List<JsonValue>> = JsonMissing.of()
        private var additionalProperties: MutableMap<String, JsonValue> = mutableMapOf()

        @JvmSynthetic
        internal fun from(folder: Folder) = apply {
            this.id = folder.id
            this.remoteId = folder.remoteId
            this.name = folder.name
            this.folderUrl = folder.folderUrl
            this.size = folder.size
            this.description = folder.description
            this.parentFolder = folder.parentFolder
            this.drive = folder.drive
            this.permissions = folder.permissions
            this.remoteCreatedAt = folder.remoteCreatedAt
            this.remoteUpdatedAt = folder.remoteUpdatedAt
            this.fieldMappings = folder.fieldMappings
            this.modifiedAt = folder.modifiedAt
            this.remoteData = folder.remoteData
            additionalProperties(folder.additionalProperties)
        }

        fun id(id: String) = id(JsonField.of(id))

        @JsonProperty("id") @ExcludeMissing fun id(id: JsonField<String>) = apply { this.id = id }

        /** The third-party API ID of the matching object. */
        fun remoteId(remoteId: String) = remoteId(JsonField.of(remoteId))

        /** The third-party API ID of the matching object. */
        @JsonProperty("remote_id")
        @ExcludeMissing
        fun remoteId(remoteId: JsonField<String>) = apply { this.remoteId = remoteId }

        /** The folder's name. */
        fun name(name: String) = name(JsonField.of(name))

        /** The folder's name. */
        @JsonProperty("name")
        @ExcludeMissing
        fun name(name: JsonField<String>) = apply { this.name = name }

        /** The URL to access the folder. */
        fun folderUrl(folderUrl: String) = folderUrl(JsonField.of(folderUrl))

        /** The URL to access the folder. */
        @JsonProperty("folder_url")
        @ExcludeMissing
        fun folderUrl(folderUrl: JsonField<String>) = apply { this.folderUrl = folderUrl }

        /** The folder's size, in bytes. */
        fun size(size: Long) = size(JsonField.of(size))

        /** The folder's size, in bytes. */
        @JsonProperty("size")
        @ExcludeMissing
        fun size(size: JsonField<Long>) = apply { this.size = size }

        /** The folder's description. */
        fun description(description: String) = description(JsonField.of(description))

        /** The folder's description. */
        @JsonProperty("description")
        @ExcludeMissing
        fun description(description: JsonField<String>) = apply { this.description = description }

        /** The folder that the folder belongs to. */
        fun parentFolder(parentFolder: String) = parentFolder(JsonField.of(parentFolder))

        /** The folder that the folder belongs to. */
        @JsonProperty("parent_folder")
        @ExcludeMissing
        fun parentFolder(parentFolder: JsonField<String>) = apply {
            this.parentFolder = parentFolder
        }

        /** The drive that the folder belongs to. */
        fun drive(drive: String) = drive(JsonField.of(drive))

        /** The drive that the folder belongs to. */
        @JsonProperty("drive")
        @ExcludeMissing
        fun drive(drive: JsonField<String>) = apply { this.drive = drive }

        /**
         * The Permission object is used to represent a user's or group's access to a File or
         * Folder. Permissions are unexpanded by default. Use the query param `expand=permissions`
         * to see more details under `GET /folders`.
         */
        fun permissions(permissions: List<String>) = permissions(JsonField.of(permissions))

        /**
         * The Permission object is used to represent a user's or group's access to a File or
         * Folder. Permissions are unexpanded by default. Use the query param `expand=permissions`
         * to see more details under `GET /folders`.
         */
        @JsonProperty("permissions")
        @ExcludeMissing
        fun permissions(permissions: JsonField<List<String>>) = apply {
            this.permissions = permissions
        }

        /** When the third party's folder was created. */
        fun remoteCreatedAt(remoteCreatedAt: OffsetDateTime) =
            remoteCreatedAt(JsonField.of(remoteCreatedAt))

        /** When the third party's folder was created. */
        @JsonProperty("remote_created_at")
        @ExcludeMissing
        fun remoteCreatedAt(remoteCreatedAt: JsonField<OffsetDateTime>) = apply {
            this.remoteCreatedAt = remoteCreatedAt
        }

        /** When the third party's folder was updated. */
        fun remoteUpdatedAt(remoteUpdatedAt: OffsetDateTime) =
            remoteUpdatedAt(JsonField.of(remoteUpdatedAt))

        /** When the third party's folder was updated. */
        @JsonProperty("remote_updated_at")
        @ExcludeMissing
        fun remoteUpdatedAt(remoteUpdatedAt: JsonField<OffsetDateTime>) = apply {
            this.remoteUpdatedAt = remoteUpdatedAt
        }

        @JsonProperty("field_mappings")
        @ExcludeMissing
        fun fieldMappings(fieldMappings: JsonValue) = apply { this.fieldMappings = fieldMappings }

        /** This is the datetime that this object was last updated by Merge */
        fun modifiedAt(modifiedAt: OffsetDateTime) = modifiedAt(JsonField.of(modifiedAt))

        /** This is the datetime that this object was last updated by Merge */
        @JsonProperty("modified_at")
        @ExcludeMissing
        fun modifiedAt(modifiedAt: JsonField<OffsetDateTime>) = apply {
            this.modifiedAt = modifiedAt
        }

        fun remoteData(remoteData: List<JsonValue>) = remoteData(JsonField.of(remoteData))

        @JsonProperty("remote_data")
        @ExcludeMissing
        fun remoteData(remoteData: JsonField<List<JsonValue>>) = apply {
            this.remoteData = remoteData
        }

        fun additionalProperties(additionalProperties: Map<String, JsonValue>) = apply {
            this.additionalProperties.clear()
            this.additionalProperties.putAll(additionalProperties)
        }

        @JsonAnySetter
        fun putAdditionalProperty(key: String, value: JsonValue) = apply {
            this.additionalProperties.put(key, value)
        }

        fun putAllAdditionalProperties(additionalProperties: Map<String, JsonValue>) = apply {
            this.additionalProperties.putAll(additionalProperties)
        }

        fun build(): Folder =
            Folder(
                id,
                remoteId,
                name,
                folderUrl,
                size,
                description,
                parentFolder,
                drive,
                permissions.map { it.toUnmodifiable() },
                remoteCreatedAt,
                remoteUpdatedAt,
                fieldMappings,
                modifiedAt,
                remoteData.map { it.toUnmodifiable() },
                additionalProperties.toUnmodifiable(),
            )
    }
}
